editoast/src/views/infra/delimited_area.rs

```diff
@@ -0,0 +1,934 @@
+            });
+    let (valid_exits, invalid_exits): (Vec<_>, Vec<_>) =
+        exits
+            .into_iter()
+            .partition_map(|exit| match validate_location(&exit, &infra_cache) {
+                Ok(_) => Either::Left(exit),
+                Err(e) => Either::Right((exit, e)),
+            });
+
+    if !(invalid_exits.is_empty() && invalid_entries.is_empty()) {
+        let invalid_locations = invalid_entries
+            .into_iter()
+            .chain(invalid_exits.into_iter())
+            .collect::<Vec<_>>();
+        return Err(DelimitedAreaError::InvalidLocations { invalid_locations }.into());
+    }
+
+    // Retrieve the track ranges
+
+    Ok(Json(DelimitedAreaResponse {
+        track_ranges: track_ranges_from_locations(valid_entries, valid_exits, &graph, &infra_cache),
+    }))
+}
+
+/// Check whether a location is valid on a given infra cache, i.e. if it matches a track on the infra
+/// and if it is within bounds.
+fn validate_location(
+    location: &DirectedLocation,
+    infra_cache: &InfraCache,
+) -> StdResult<(), InputError> {
+    // Check if the location track exists on the infra
+
+    let track_length = infra_cache
+        .track_sections()
+        .get(&location.track.0)
+        .ok_or(InputError::TrackDoesNotExist(location.track.0.clone()))?
+        .unwrap_track_section()
+        .length;
+
+    // Check if the location is within bounds on the track
+
+    if location.position < 0. || track_length < location.position {
+        Err(InputError::LocationOutOfBounds {
+            track: location.track.0.clone(),
+            position: location.position,
+            track_length,
+        })
+    } else {
+        Ok(())
+    }
+}
+
+fn track_ranges_from_locations(
+    entries: Vec<DirectedLocation>,
+    exits: Vec<DirectedLocation>,
+    graph: &Graph,
+    infra_cache: &InfraCache,
+) -> Vec<DirectionalTrackRange> {
+    entries
+        .iter()
+        .flat_map(|entry| {
+            impacted_tracks(
+                entry,
+                exits.iter().collect::<Vec<&DirectedLocation>>(),
+                graph,
+                infra_cache,
+                MAXIMUM_DISTANCE,
+            )
+        })
+        .collect::<Vec<_>>()
+}
+
+fn impacted_tracks(
+    entry: &DirectedLocation,
+    exits: Vec<&DirectedLocation>,
+    graph: &Graph,
+    infra_cache: &InfraCache,
+    max_distance: f64,
+) -> Vec<DirectionalTrackRange> {
+    // Map track identifiers to their list of associated exits:
+    let exits = {
+        let mut tracks_to_exits: HashMap<&Identifier, Vec<&DirectedLocation>> = HashMap::new();
+        exits.into_iter().for_each(|exit| {
+            let track_id = &exit.track;
+            if let Some(track_exits) = tracks_to_exits.get_mut(track_id) {
+                track_exits.push(exit);
+            } else {
+                let track_exits = vec![exit];
+                tracks_to_exits.insert(track_id, track_exits);
+            };
+        });
```

Comment:
[`entry`](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry) may be useful here:

```suggestion
		for exit in exits {
            tracks_to_exits.entry(&exit.track).or_default().push(exit);
        }
```